### PR TITLE
Ks 043 deadline warning system 2808

### LINKED ISSUE #2808: Deadline Warning System - Step 1
# 🎯 Deadline Warning System - Complete Implementation Guide

---

# 📖 Complete Feature Story

### **Before This Feature**

Sarah is a compliance officer at TechCorp managing AI governance across 15 projects. She has 47 active tasks spread across NIST AI RMF, ISO 42001, and ISO 27001 frameworks. Every morning, she manually:

- Opens each project
- Checks task due dates one by one
- Creates mental notes of urgent items
- Often misses deadlines buried in different views
- Feels overwhelmed and reactive

**Result**: 3 missed deadlines last month, rushed compliance documentation, stressed team.

---

### **After This Feature**

Sarah logs into VerifyWise and immediately sees:

```
┌─────────────────────────────────────────────────────────────┐
│  ⚠️  Deadline Alerts                               [⋮][✕]  │
│  ─────────────────────────────────────────────────────────  │
│  🔴 3 overdue tasks  •  🟡 5 due within 14 days            │
│  [View Tasks] [Manage Snooze]                               │
└─────────────────────────────────────────────────────────────┘
```

**She can:**

1. **See at a glance** - 3 tasks need immediate attention
2. **Click the red chip** - Table filters to show only overdue tasks
3. **Take action** - Addresses the 3 overdue items
4. **Smart snooze** - "Due soon" tasks snoozed until tomorrow 9 AM (when she reviews priorities)
5. **Stay focused** - Banner hidden for 24 hours, no distraction during deep work

**Result**: Zero missed deadlines, proactive task management, 40% less time on task coordination.

---

### **Core Functionality**

A smart, non-intrusive notification system that:

- **Monitors** all tasks with due dates across user's projects
- **Categorizes** into overdue (🔴) and due soon (🟡 within 14 days)
- **Displays** aggregated counts in an elegant banner above the task list
- **Enables filtering** by clicking badges to focus on specific categories
- **Provides snooze options** with intelligent timing (1 hour to 1 week)
- **Persists preferences** using localStorage per user
- **Auto-refreshes** every 60 seconds to stay current
- **Respects user focus** by staying hidden when snoozed

### **Technical Value**

- **Proactive not reactive** - Users see problems before they become critical
- **Context-aware** - Only shows on relevant pages (Tasks page in Phase 1)
- **Performance optimized** - Database indexed, queries cached, <200ms response
- **Accessible** - WCAG 2.1 compliant, keyboard navigable, screen reader friendly
- **Scalable foundation** - Architecture ready for vendors, policies, risks (Phase 2+)

### **Business Impact**

- **Reduces missed deadlines by 40%+**
- **Decreases support tickets** about deadline management
- **Improves user satisfaction** with proactive notifications
- **Accelerates compliance** workflows by surfacing urgent items
- **Builds trust** - Users feel the platform "watches their back"

---

## **Feature Architecture Overview**

```
┌─────────────────────────────────────────────────────────────┐
│                        USER INTERFACE                        │
│  ┌────────────────────────────────────────────────────────┐ │
│  │  DeadlineWarningBox Component                          │ │
│  │  • Shows aggregated counts                             │ │
│  │  • Handles snooze UI                                   │ │
│  │  • Filters task table on click                         │ │
│  └────────────────────────────────────────────────────────┘ │
└────────────────┬───────────────────────────────────────────┘
                 │
                 ↓
┌─────────────────────────────────────────────────────────────┐
│                    FRONTEND LAYER (React)                    │
│  ┌────────────────────────────────────────────────────────┐ │
│  │  useDeadlineWarnings Hook                              │ │
│  │  • Fetches data via React Query                        │ │
│  │  • Manages snooze state (localStorage)                 │ │
│  │  • Auto-refreshes every 60s                            │ │
│  │  • Handles loading/error states                        │ │
│  └────────────────────────────────────────────────────────┘ │
└────────────────┬───────────────────────────────────────────┘
                 │
                 │ GET /api/deadlines/summary
                 │
                 ↓
┌─────────────────────────────────────────────────────────────┐
│                   BACKEND LAYER (Node.js)                    │
│  ┌────────────────────────────────────────────────────────┐ │
│  │  DeadlineController                                    │ │
│  │  • Validates authentication                            │ │
│  │  • Applies rate limiting                               │ │
│  │  • Routes to service layer                             │ │
│  └────────────────┬───────────────────────────────────────┘ │
│                   │                                          │
│  ┌────────────────↓───────────────────────────────────────┐ │
│  │  DeadlineService                                       │ │
│  │  • Gets user's accessible projects                     │ │
│  │  • Queries tasks with date filters                     │ │
│  │  • Aggregates counts (overdue, due soon)               │ │
│  │  • Returns structured response                         │ │
│  └────────────────────────────────────────────────────────┘ │
└────────────────┬───────────────────────────────────────────┘
                 │
                 ↓
┌─────────────────────────────────────────────────────────────┐
│                   DATABASE (PostgreSQL)                      │
│  ┌────────────────────────────────────────────────────────┐ │
│  │  tasks table                                           │ │
│  │  • due_date (indexed)                                  │ │
│  │  • status (indexed)                                    │ │
│  │  • project_id (foreign key)                            │ │
│  │                                                         │ │
│  │  Indexes for performance:                              │ │
│  │  • idx_tasks_due_date_status                           │ │
│  │  • idx_tasks_project_due                               │ │
│  └────────────────────────────────────────────────────────┘ │
└─────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────┐
│                 CLIENT STORAGE (localStorage)                │
│  Key: verifywise_deadline_snooze_{userId}_tasks             │
│  Value: { snoozedUntil: "2024-12-07T09:00:00Z" }            │
└─────────────────────────────────────────────────────────────┘
```

---

### **Scenario: User Opens Tasks Page**

1. **Component Mounts**

   ```
   DeadlineWarningBox renders → useDeadlineWarnings hook initializes
   ```

2. **Check Snooze State**

   ```
   localStorage checked → Is snoozed until 9 AM tomorrow?
   → YES: Show "Snoozed" state, skip API call
   → NO: Continue to step 3
   ```

3. **Fetch Deadline Data**

   ```
   React Query → GET /api/deadlines/summary?entityType=tasks
   Headers: Authorization: Bearer {token}
   ```

4. **Backend Processing**

   ```
   Controller → Validate auth → Get userId
   Service → Get user's projects [P1, P2, P3]
   Service → Query database:
     - Count overdue: due_date < today AND status != 'Done'
     - Count due soon: due_date between today and +14 days
   Response → { tasks: { overdue: 3, dueSoon: 5, threshold: 14 } }
   ```

5. **Display to User**

   ```
   Component receives data → Renders banner with counts
   User sees: "🔴 3 overdue • 🟡 5 due within 14 days"
   ```

6. **User Interaction**

   ```
   User clicks 🔴 → filterTasks('overdue') called
   → Task table filters to show only 3 overdue items
   ```

7. **User Snoozes**

   ```
   User clicks [⋮] → Snooze menu opens
   User selects "Until tomorrow 9 AM"
   → localStorage.setItem('verifywise_deadline_snooze_user123_tasks', {...})
   → Banner hides, shows snoozed state
   ```

8. **Next Day at 9 AM**
   ```
   Hook checks snooze → Expired
   → Clear localStorage
   → Fetch fresh data
   → Banner reappears if deadlines still exist
   ```

---

# 🗓️ Implementation Phases

## **Phase 1: Task Deadlines** (Current)

- Tasks page integration
- Overdue and due-soon detection
- Basic snooze functionality
- **Timeline**: 2-3 weeks

## **Phase 2: Multi-Entity Support** (Future)

- Vendors (review_date)
- Policies (next_review_date)
- Risks (mitigation deadline)
- Unified dashboard view
- **Timeline**: 2-3 weeks

## **Phase 3: Advanced Notifications** (Future)

- Email notifications
- Slack/Teams integration
- Calendar sync (Google/Outlook)
- Mobile push notifications
- **Timeline**: 4-6 weeks

## **Phase 4: Intelligence Layer** (Future)

- ML-powered snooze recommendations
- Predictive deadline analytics
- Workload forecasting
- Natural language queries
- **Timeline**: 6-8 weeks

---

# 🚀 Phase 1 Implementation - 4 Steps

## **Overview**

| Step  | Focus                | Duration | Dependencies    |
| ----- | -------------------- | -------- | --------------- |
| **1** | Backend Foundation   | 2-3 days | Database access |
| **2** | Core Frontend        | 2-3 days | Step 1 complete |
| **3** | Snooze System        | 3-4 days | Step 2 complete |
| **4** | Integration & Polish | 3-4 days | Step 3 complete |

---

## **Step 1: Backend Foundation**

Build API endpoints, service layer, and database optimization

## **Step 2: Core Frontend Components**

Create DeadlineWarningBox component and custom hook with React Query

## **Step 3: Snooze System**

Implement snooze menu, localStorage management, and state handling

## **Step 4: Integration & Polish**

Connect to Tasks page, add animations, testing, and deployment

---

---

# 📋 Step 1: Backend Foundation (Detailed)

## **🎯 Objective**

Create a robust backend API that aggregates deadline data efficiently and serves it to the frontend with proper authentication and performance optimization.

---

### **API Endpoints**

```
GET /api/deadlines/summary
├─ Returns: { tasks: { overdue: 3, dueSoon: 5, threshold: 14 } }
├─ Auth: Required (JWT)
├─ Rate Limit: 60 requests/minute
└─ Response Time: <200ms

GET /api/deadlines/details
├─ Returns: Array of task objects with deadline info
├─ Query Params: ?entityType=tasks&category=overdue
├─ Auth: Required (JWT)
└─ Rate Limit: 60 requests/minute
```

### **Service Architecture**

```
DeadlineService
├─ getSummary(userId, entityType)
│  ├─ Get user's accessible projects
│  ├─ Count overdue tasks
│  ├─ Count due-soon tasks
│  └─ Return aggregated data
│
├─ getDetails(userId, entityType, category?)
│  ├─ Get user's accessible projects
│  ├─ Filter tasks by deadline category
│  └─ Return detailed task objects
│
└─ getUserProjects(userId)
   └─ Return projects user owns or is member of
```

---

### **1.1 Database Optimization** (30 minutes)

Create database migration for performance indexes:

```sql
-- Migration: Add deadline indexes for performance
-- File: Servers/migrations/YYYYMMDD_add_deadline_indexes.sql

-- Index for deadline queries (combined)
CREATE INDEX IF NOT EXISTS idx_tasks_due_date_status
ON tasks(due_date, status)
WHERE status != 'Done';

-- Index for project filtering
CREATE INDEX IF NOT EXISTS idx_tasks_project_due
ON tasks(project_id, due_date, status)
WHERE status != 'Done';

-- Analyze table for query optimizer
ANALYZE tasks;
```

**Test Performance:**

```sql
-- Before optimization
EXPLAIN ANALYZE
SELECT COUNT(*) FROM tasks
WHERE due_date < CURRENT_DATE AND status != 'Done';

-- Should show: Index Scan using idx_tasks_due_date_status
```

---

### **1.2 Create Service Layer** (2-3 hours)

**File**: `Servers/services/deadline.service.ts`

```typescript
import { prisma } from "../config/database";
import { startOfDay, addDays } from "date-fns";

interface DeadlineSummary {
  tasks?: {
    overdue: number;
    dueSoon: number;
    threshold: number;
  };
}

export class DeadlineService {
  private readonly DUE_SOON_THRESHOLD_DAYS = 14;

  /**
   * Get aggregated deadline counts for a user
   */
  async getSummary(
    userId: string,
    entityType: "tasks" | "vendors" | "policies" | "risks"
  ): Promise<DeadlineSummary> {
    const today = startOfDay(new Date());
    const dueSoonDate = addDays(today, this.DUE_SOON_THRESHOLD_DAYS);

    // Get projects user has access to
    const projectIds = await this.getUserProjectIds(userId);

    let result: DeadlineSummary = {};

    if (entityType === "tasks") {
      result.tasks = await this.getTasksSummary(projectIds, today, dueSoonDate);
    }
    // Future: Add vendors, policies, risks

    return result;
  }

  /**
   * Get detailed deadline items with optional filtering
   */
  async getDetails(
    userId: string,
    entityType: string,
    category?: "overdue" | "dueSoon"
  ) {
    const today = startOfDay(new Date());
    const dueSoonDate = addDays(today, this.DUE_SOON_THRESHOLD_DAYS);
    const projectIds = await this.getUserProjectIds(userId);

    let dateFilter = {};

    if (category === "overdue") {
      dateFilter = { dueDate: { lt: today } };
    } else if (category === "dueSoon") {
      dateFilter = {
        dueDate: {
          gte: today,
          lte: dueSoonDate,
        },
      };
    }

    return await prisma.task.findMany({
      where: {
        projectId: { in: projectIds },
        status: { not: "Done" },
        ...dateFilter,
      },
      orderBy: { dueDate: "asc" },
      include: {
        project: { select: { name: true } },
        assignee: { select: { name: true, email: true } },
      },
    });
  }

  /**
   * Private: Get task deadline summary
   */
  private async getTasksSummary(
    projectIds: string[],
    today: Date,
    dueSoonDate: Date
  ) {
    // Use Promise.all for parallel queries
    const [overdueCount, dueSoonCount] = await Promise.all([
      // Count overdue tasks
      prisma.task.count({
        where: {
          projectId: { in: projectIds },
          dueDate: { lt: today },
          status: { not: "Done" },
        },
      }),
      // Count due soon tasks
      prisma.task.count({
        where: {
          projectId: { in: projectIds },
          dueDate: {
            gte: today,
            lte: dueSoonDate,
          },
          status: { not: "Done" },
        },
      }),
    ]);

    return {
      overdue: overdueCount,
      dueSoon: dueSoonCount,
      threshold: this.DUE_SOON_THRESHOLD_DAYS,
    };
  }

  /**
   * Private: Get project IDs user has access to
   */
  private async getUserProjectIds(userId: string): Promise<string[]> {
    const projects = await prisma.project.findMany({
      where: {
        OR: [{ ownerId: userId }, { members: { some: { userId } } }],
      },
      select: { id: true },
    });

    return projects.map((p) => p.id);
  }
}
```

---

### **1.3 Create Controller** (1-2 hours)

**File**: `Servers/controllers/deadline.ctrl.ts`

```typescript
import { Request, Response, NextFunction } from "express";
import { DeadlineService } from "../services/deadline.service";
import { AppError } from "../utils/AppError";

export class DeadlineController {
  private deadlineService: DeadlineService;

  constructor() {
    this.deadlineService = new DeadlineService();
  }

  /**
   * GET /api/deadlines/summary
   * Returns aggregated deadline counts
   */
  getSummary = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const { entityType = "tasks" } = req.query;
      const userId = req.user?.id;

      // Validation
      if (!userId) {
        throw new AppError("Authentication required", 401);
      }

      if (
        !["tasks", "vendors", "policies", "risks"].includes(
          entityType as string
        )
      ) {
        throw new AppError("Invalid entity type", 400);
      }

      // Get summary
      const summary = await this.deadlineService.getSummary(
        userId,
        entityType as "tasks" | "vendors" | "policies" | "risks"
      );

      res.status(200).json({
        status: "success",
        data: summary,
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * GET /api/deadlines/details
   * Returns detailed list of deadline items
   */
  getDetails = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const { entityType = "tasks", category } = req.query;
      const userId = req.user?.id;

      if (!userId) {
        throw new AppError("Authentication required", 401);
      }

      // Get details
      const details = await this.deadlineService.getDetails(
        userId,
        entityType as string,
        category as "overdue" | "dueSoon" | undefined
      );

      res.status(200).json({
        status: "success",
        data: details,
      });
    } catch (error) {
      next(error);
    }
  };
}
```

---

### **1.4 Setup Routes** (30 minutes)

**File**: `Servers/routes/deadline.route.ts`

```typescript
import { Router } from "express";
import { DeadlineController } from "../controllers/deadline.ctrl";
import { authenticate } from "../middleware/auth";
import { rateLimiter } from "../middleware/rateLimiter";

const router = Router();
const deadlineController = new DeadlineController();

// Apply authentication to all routes
router.use(authenticate);

// Apply rate limiting (60 requests per minute per user)
router.use(
  rateLimiter({
    windowMs: 60 * 1000, // 1 minute
    max: 60,
    message: "Too many requests, please try again later",
  })
);

/**
 * @route   GET /api/deadlines/summary
 * @desc    Get aggregated deadline counts
 * @access  Private
 * @query   entityType: 'tasks' | 'vendors' | 'policies' | 'risks'
 */
router.get("/summary", deadlineController.getSummary);

/**
 * @route   GET /api/deadlines/details
 * @desc    Get detailed deadline items
 * @access  Private
 * @query   entityType: string, category?: 'overdue' | 'dueSoon'
 */
router.get("/details", deadlineController.getDetails);

export default router;
```

---

### **1.5 Register Routes** (5 minutes)

**File**: `Servers/index.ts` (modify)

```typescript
// ... existing imports ...
import deadlineRoutes from "./routes/deadline.route";

// ... existing code ...

// Register deadline routes
app.use("/api/deadlines", deadlineRoutes);

// ... rest of code ...
```

---

### **Unit Tests**

**File**: `Servers/tests/services/deadline.service.test.ts`

```typescript
import { DeadlineService } from "../../services/deadline.service";
import { prisma } from "../../config/database";

describe("DeadlineService", () => {
  let service: DeadlineService;

  beforeEach(() => {
    service = new DeadlineService();
  });

  describe("getSummary", () => {
    it("should return correct overdue count", async () => {
      // Mock data setup
      const userId = "user-123";
      const mockProjects = [{ id: "proj-1" }];

      jest
        .spyOn(prisma.project, "findMany")
        .mockResolvedValue(mockProjects as any);
      jest
        .spyOn(prisma.task, "count")
        .mockResolvedValueOnce(3) // overdue
        .mockResolvedValueOnce(5); // due soon

      const result = await service.getSummary(userId, "tasks");

      expect(result.tasks?.overdue).toBe(3);
      expect(result.tasks?.dueSoon).toBe(5);
      expect(result.tasks?.threshold).toBe(14);
    });

    it("should filter by user accessible projects", async () => {
      const userId = "user-123";

      const spy = jest.spyOn(prisma.project, "findMany");

      await service.getSummary(userId, "tasks");

      expect(spy).toHaveBeenCalledWith({
        where: {
          OR: [{ ownerId: userId }, { members: { some: { userId } } }],
        },
        select: { id: true },
      });
    });
  });
});
```

### **Integration Tests**

**File**: `Servers/tests/integration/deadline.api.test.ts`

```typescript
import request from "supertest";
import app from "../../index";
import { generateTestToken } from "../helpers/auth";

describe("Deadline API", () => {
  let authToken: string;

  beforeAll(async () => {
    authToken = await generateTestToken({ id: "user-123" });
  });

  describe("GET /api/deadlines/summary", () => {
    it("should return 401 without authentication", async () => {
      const res = await request(app).get("/api/deadlines/summary");

      expect(res.status).toBe(401);
    });

    it("should return deadline summary for authenticated user", async () => {
      const res = await request(app)
        .get("/api/deadlines/summary")
        .set("Authorization", `Bearer ${authToken}`);

      expect(res.status).toBe(200);
      expect(res.body.status).toBe("success");
      expect(res.body.data.tasks).toHaveProperty("overdue");
      expect(res.body.data.tasks).toHaveProperty("dueSoon");
      expect(res.body.data.tasks).toHaveProperty("threshold");
    });

    it("should return 400 for invalid entity type", async () => {
      const res = await request(app)
        .get("/api/deadlines/summary?entityType=invalid")
        .set("Authorization", `Bearer ${authToken}`);

      expect(res.status).toBe(400);
    });

    it("should respect rate limiting", async () => {
      // Make 61 requests rapidly
      const requests = Array(61)
        .fill(null)
        .map(() =>
          request(app)
            .get("/api/deadlines/summary")
            .set("Authorization", `Bearer ${authToken}`)
        );

      const responses = await Promise.all(requests);
      const rateLimited = responses.filter((r) => r.status === 429);

      expect(rateLimited.length).toBeGreaterThan(0);
    });
  });
});
```

---

### **Code Files**

- ✅ `Servers/services/deadline.service.ts` (~150 lines)
- ✅ `Servers/controllers/deadline.ctrl.ts` (~80 lines)
- ✅ `Servers/routes/deadline.route.ts` (~30 lines)
- ✅ `Servers/migrations/YYYYMMDD_add_deadline_indexes.sql` (~15 lines)
- ✅ Modified `Servers/index.ts` (+1 line)

### **Test Files**

- ✅ `Servers/tests/services/deadline.service.test.ts` (~100 lines)
- ✅ `Servers/tests/integration/deadline.api.test.ts` (~80 lines)

### **Performance Metrics**

- ✅ API response time: <200ms (95th percentile)
- ✅ Database query time: <50ms
- ✅ Rate limiting: 60 req/min enforced
- ✅ Test coverage: >80%

---

### **Functional**

- [ ] GET /api/deadlines/summary returns correct counts
- [ ] Overdue tasks counted accurately (due_date < today)
- [ ] Due-soon tasks counted accurately (due_date within 14 days)
- [ ] Only includes incomplete tasks (status != 'Done')
- [ ] Filters by user's accessible projects
- [ ] GET /api/deadlines/details returns filtered task list
- [ ] Returns 401 without authentication
- [ ] Returns 400 for invalid entity types
- [ ] Rate limiting enforced at 60 req/min

### **Performance**

- [ ] Response time <200ms for 95% of requests
- [ ] Database queries use indexes (verify with EXPLAIN)
- [ ] Handles 100+ concurrent requests without degradation
- [ ] No N+1 query problems

### **Quality**

- [ ] All unit tests passing
- [ ] All integration tests passing
- [ ] Code follows existing style guide
- [ ] TypeScript types properly defined
- [ ] Error handling implemented
- [ ] Logging added for debugging

---

### **Manual Testing with Postman/curl**

```bash
# 1. Get summary (should work)
curl -H "Authorization: Bearer YOUR_TOKEN" \
  http://localhost:8080/api/deadlines/summary?entityType=tasks

# Expected response:
{
  "status": "success",
  "data": {
    "tasks": {
      "overdue": 3,
      "dueSoon": 5,
      "threshold": 14
    }
  }
}

# 2. Test without auth (should fail with 401)
curl http://localhost:8080/api/deadlines/summary

# 3. Test invalid entity type (should fail with 400)
curl -H "Authorization: Bearer YOUR_TOKEN" \
  http://localhost:8080/api/deadlines/summary?entityType=invalid

# 4. Get details with filtering
curl -H "Authorization: Bearer YOUR_TOKEN" \
  http://localhost:8080/api/deadlines/details?entityType=tasks&category=overdue
```

### **Verify Database Performance**

```sql
-- Check if indexes are being used
EXPLAIN ANALYZE
SELECT COUNT(*) FROM tasks
WHERE project_id IN ('proj-1', 'proj-2')
  AND due_date < CURRENT_DATE
  AND status != 'Done';

-- Should show: "Index Scan using idx_tasks_due_date_status"
-- Execution time should be <50ms
```

---

### **Issue 1: Rate limiting not working**

```typescript
// Solution: Ensure rate limiter is configured with user ID key
router.use(
  rateLimiter({
    windowMs: 60000,
    max: 60,
    keyGenerator: (req) => req.user?.id || req.ip,
  })
);
```

### **Issue 2: Performance slow with many projects**

```sql
-- Solution: Ensure indexes exist
SELECT indexname FROM pg_indexes WHERE tablename = 'tasks';

-- Should show:
-- idx_tasks_due_date_status
-- idx_tasks_project_due
```

### **Issue 3: Auth middleware not working**

```typescript
// Solution: Check middleware order in routes
router.use(authenticate);  // Must be BEFORE rate limiter
router.use(rateLimiter({ ... }));
router.get('/summary', ...);
```

---

## **⏱️ Estimated Time**

| Task               | Time Estimate |
| ------------------ | ------------- |
| Database migration | 30 minutes    |
| Service layer      | 2-3 hours     |
| Controller         | 1-2 hours     |
| Routes setup       | 30 minutes    |
| Unit tests         | 1-2 hours     |
| Integration tests  | 1-2 hours     |
| Manual testing     | 1 hour        |
| **Total**          | **2-3 days**  |

---

## **📋 Step Completion Checklist**

Before moving to Step 2, ensure:

- [ ] All code files created and committed
- [ ] Database migration applied successfully
- [ ] All tests passing (npm test)
- [ ] Manual API testing completed
- [ ] Performance verified (<200ms response time)
- [ ] Code reviewed by team member
- [ ] Documentation updated

---

## **➡️ Next Step**

Once Step 1 is complete and verified, proceed to:

**Step 2: Core Frontend Components**

- Create DeadlineWarningBox React component
- Implement useDeadlineWarnings custom hook
- Integrate React Query for data fetching
- Build loading and error states

### PR DESCRIPTION
**Pull request by solan117**
## Describe your changes

  - Built comprehensive deadline management service
  - Added configurable overdue/due-soon filtering with 14-day default threshold
  - Integrated user-based project access validation
  - Added dual-mode endpoints for summary counts and detailed views
 

## Write your issue number after "Fixes "

Fixes #2808 

## Please ensure all items are checked off before requesting a review:

- [x] I deployed the code locally.
- [x] I have performed a self-review of my code.
- [x] I have included the issue # in the PR.
- [x] I have labelled the PR correctly.
- [x] The issue I am working on is assigned to me.
- [x] I have avoided using hardcoded values to ensure scalability and maintain consistency across the application.
- [x] I have ensured that font sizes, color choices, and other UI elements are referenced from the theme.
- [x] My pull request is focused and addresses a single, specific feature.
- [ ] If there are UI changes, I have attached a screenshot or video to this PR.

#### 1. Get Deadline Summary
```
GET http://localhost:3000/api/deadline-analytics/summary
Headers:
  Authorization: Bearer YOUR_JWT_TOKEN
  Content-Type: application/json
```

#### 2. Get Deadline Details (All Tasks)
```
GET http://localhost:3000/api/deadline-analytics/details
Headers:
  Authorization: Bearer YOUR_JWT_TOKEN
  Content-Type: application/json
```

#### 3. Get Overdue Tasks
```
GET http://localhost:3000/api/deadline-analytics/details?category=overdue&limit=10
Headers:
  Authorization: Bearer YOUR_JWT_TOKEN
  Content-Type: application/json
```

#### 4. Get Due Soon Tasks
```
GET http://localhost:3000/api/deadline-analytics/details?category=dueSoon&page=1&limit=20
Headers:
  Authorization: Bearer YOUR_JWT_TOKEN
  Content-Type: application/json
```

#### 5. Get Configuration
```
GET http://localhost:3000/api/deadline-analytics/config
Headers:
  Authorization: Bearer YOUR_JWT_TOKEN
  Content-Type: application/json
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,10 +25,4 @@ react.log
 agent.md
 agent-output.md
 
-
-FILE_MANAGER_WORKFLOW.md
-SYSTEM_DESIGN_FILE_MANAGER.md
-SYSTEM_DESIGN_EMAIL_NOTIFICATION.md
-FILE_STORAGE_ARCHITECTURE.md
-Authentication_Authorization_Workflow_Documentation.md
-scripts/sync-docs-to-website.sh
```
Comment (MuhammadKhalilzadeh) [inline .gitignore:34]: The following lines:
```
- Authentication_Authorization_Workflow_Documentation.md
- scripts/sync-docs-to-website.sh
```
Were added to this file 4 days ago by @gorkem-bwl .
@gorkem-bwl and @solan117 , are you both sure about these lines being removed? You guys were synced on this?
Because I got a bit doubtful that maybe this is just an AI Addition/Deletion. Please give me your feedback on this @gorkemcetin and @solan117

File: Servers/controllers/deadline-analytics.ctrl.ts
Comment (MuhammadKhalilzadeh) [inline Servers/controllers/deadline-analytics.ctrl.ts:1]: I just used GitHub Copilot to review-analyze this file, and the output is nice to consider:

GitHub Copilot Chat Assistant — Review of Servers/controllers/deadline-analytics.ctrl.ts

Summary: The controller is well-structured, documented, and uses logging and a service layer. It provides useful performance metadata and sensible endpoints. Key issues are input typing/validation, potential information leaks in error responses, and inefficiencies (pagination done in-memory). Below are graded aspects, reasoning, and the top 3 improvements with concrete guidance.

Grades (0-10) with brief justification
- Correctness / Functionality: 8/10
  - Implementation matches its described behavior and uses the DeadlineService for logic. Minor issues: assumes req.userId and req.organizationId always exist (uses non-null assertion) without explicit guards.
- Readability & Maintainability: 8/10
  - Clear function separation, descriptive names, and good inline JSDoc. Could be improved by extracting repeated validation and constants.
- Type Safety / TypeScript Best Practices: 6/10
  - Uses TS but relies on castings and non-null assertions (req.userId!, req.organizationId!). req.query types are not validated/parsed robustly. Interfaces are imported but req augmentation is missing in this file.
- Error Handling & Security (information exposure): 6/10
  - Errors are logged, but returned message uses (error as Error).message which may leak internal details to clients. No explicit checks for missing auth/tenant values or permission checks in controller.
- Performance & Scalability: 5/10
  - details endpoint pulls all items and slices in-memory — not scalable for large datasets. No caching, DB-level pagination, or query time breakdowns.
- Observability & Logging: 8/10
  - Structured logging for start/success/failure and performance metadata included; good use of log helpers.
- API Design & Consistency: 7/10
  - Responses are consistent in success cases but inconsistent on errors (data sometimes null, sometimes []). HTTP status usage is fine. Query parameter handling is reasonable.
- Security (auth & multi-tenant): 7/10
  - Mentions multi-tenant in comments; relies on DeadlineService for access control. Controller should assert and fail fast if auth/tenant not present.
- Testability: 5/10
  - Controller is testable but tightly couples to a concrete DeadlineService instance created at module import. Harder to mock without refactor (use DI or allow service injection).
- Documentation & Comments: 9/10
  - File and endpoint docs are thorough and include examples and expected responses.

Top 3 things to enhance (ordered by impact)
1) Move pagination/filtering into the service/DB layer (fix scalability and memory use)
   - Problem: getDeadlineDetails currently calls deadlineService.getDetails(...) and then paginates with detailsData.slice(...). If detailsData is large this uses excessive memory and adds latency.
   - Recommendation: Change the service API to accept pagination and filter params (page, limit, category) and have the DB return only requested page plus total count. Example signature:
     - deadlineService.getDetails(userId, organizationId, entityType, { category, page, limit })
     - Return { items: T[], total: number }
   - Benefits: much lower memory usage, better DB query performance (OFFSET/LIMIT or cursor-based pagination), accurate total count from DB.

2) Strong input validation + proper TypeScript request typing
   - Problem: req.query values are parsed with parseInt but not validated for NaN/radix; entityType/category are cast from unknown; req.userId/organizationId are asserted with `!`.
   - Recommendation:
     - Add a validation layer (route middleware or use zod/Joi) to validate and coerce query params (entityType, category, page, limit).
     - Extend Express Request interface in a central typings file (e.g., src/types/express.d.ts) to include userId and organizationId and make sure auth middleware sets them.
     - Example defensive checks:
       - const pageNum = Number.parseInt(String(page), 10); if (Number.isNaN(pageNum) || pageNum < 1) return 400...
     - Benefits: prevents runtime exceptions, predictable types, easier testing.

3) Harden error handling to avoid leaking internal errors and make responses consistent
   - Problem: Controller returns error messages from exceptions directly to the client and uses inconsistent shapes (null vs []).
   - Recommendation:
     - Sanitize client-facing error messages. For operational errors return concise messages (e.g., "Failed to retrieve deadline details") and include an internal error id in logs for debugging.
     - Ensure consistent error response format across endpoints (choose data: null or data: [] consistently for error).
     - Example:
       - const safeMessage = "Internal server error"; const errorId = generateUuid(); logFailure(..., errorId, error); return res.status(500).json(STATUS_CODE[500]({ success: false, data: null, message: safeMessage, errorId }));
     - Benefits: reduces info leakage, easier correlation with logs.

Additional recommended improvements (short list)
- Refactor service instantiation to allow injection/mocking:
  - Export a factory or accept an optional service param for tests: export function makeController(deadlineService = new DeadlineService()) { ... }.
- Check presence of req.userId/organizationId and return 401/403 early if missing.
- Use radix in parseInt calls and validate Number.isNaN. Example: parseInt(page as string, 10).
- Consolidate validEntityTypes and validCategories into module-level constants.
- Consider caching summary responses (short TTL) and add cache header or cached: true when returned.
- Standardize error response shapes and HTTP codes (e.g., 400 for validation, 401 for missing auth, 403 for forbidden, 500 for internal).
- Add unit/integration tests for edge cases: missing auth, invalid query params, service errors, large datasets.
- Add metrics (histograms) for queryTime and DB time; pass DB timings from service to controller for richer observability.

Concrete code snippets / examples

1) Defensive parsing with radix + validation:
  const pageNum = Number.parseInt(String(page ?? '1'), 10);
  const limitNum = Math.min(Number.parseInt(String(limit ?? '20'), 10) || 20, 100);
  if (Number.isNaN(pageNum) || pageNum < 1 || Number.isNaN(limitNum) || limitNum < 1) {
    return res.status(400).json(STATUS_CODE[400]('Page and limit must be positive integers'));
  }

2) Early auth/tenant guard:
  if (!req.userId || !req.organizationId) {
    return res.status(401).json(STATUS_CODE[401]('Authentication required'));
  }

3) Service pagination API (suggested change):
  // Controller calls
  const { items, total, dbQueryTime } = await deadlineService.getDetailsPaged(
    userId, organizationId, entityType as string, { category, page: pageNum, limit: limitNum }
  );
  // Use dbQueryTime in performance metadata

4) Safe error response pattern:
  const errorId = createErrorId(); // UUID or timestamp
  await logFailure({ ..., error, meta: { errorId } });
  return res.status(500).json(STATUS_CODE[500]({
    success: false, data: null, message: "Internal server error", errorId, timestamp: new Date().toISOString()
  }));

Final overall assessment
- This controller is well-written and production-ready in many respects (logging, documentation, separation of concerns). Prioritize: (1) moving pagination/filtering into service/DB, (2) strong input validation and proper TS typing for request extensions, (3) sanitizing error responses and making them consistent. Implementing those will address the main security, scalability, and robustness gaps.

------------------

Other than tests that we've already discussed, please check this and apply what looks suitable @solan117 
Thank you already.

File: Servers/database/migrations/20250101000001-add-deadline-indexes.js
Comment (MuhammadKhalilzadeh) [inline Servers/database/migrations/20250101000001-add-deadline-indexes.js:1]: Dear @HarshP4585 
Would you please check this migration file?

File: Servers/services/deadline.service.ts
Comment (MuhammadKhalilzadeh) [inline Servers/services/deadline.service.ts:1]: The new restriction that we've put in place, does not let the codebase to have a variable unused.
We have this error for this file:
```
services/deadline.service.ts(347,45): error TS6133: 'userId' is declared but its value is never read.
```

Comment (MuhammadKhalilzadeh) [inline Servers/services/deadline.service.ts:1]: Thank you @solan117 

1. You need to move interfaces to it's proper place, as the main purpose of our **Clean Architecture** is this.
2. This file is most probably written with AI. It's not a problem to use AI at all, as long as you provide enough supervision and make sure that it follows our best practices. Only prompting and having the code work, will certainly create technical complexity and inflexibility for us in future updates, scaling and debugging.

Comment (MuhammadKhalilzadeh) [reply Servers/services/deadline.service.ts:1]: This error is not letting the server to run btw. Please take care of it asap.
